src/planner/ompl/Planner.cpp

```diff
@@ -372,6 +374,119 @@ trajectory::InterpolatedPtr planCRRTConnect(
+    double _maxDistanceBtwValidityChecks,
+    double _timeout,
+    size_t _maxEmptySteps,
+    trajectory::InterpolatedPtr _originalTraj)
+{
+  if (_timeout < 0)
+  {
+    throw std::invalid_argument("Timeout must be >= 0");
+  }
+
+  // Step 1: Generate the space information of OMPL type
+  auto si = getSpaceInformation(
+      _stateSpace,
+      _interpolator,
+      std::move(_dmetric),
+      std::move(_sampler),
+      std::move(_validityConstraint),
+      std::move(_boundsConstraint),
+      std::move(_boundsProjector),
+      _maxDistanceBtwValidityChecks);
+
+  // Step 2: Convert AIKIDO Interpolated Trajectory to OMPL path
+  auto path = toOMPLTrajectory(_originalTraj, si);
+
+  // Step 3: Use the OMPL methods to simplify the path
+  ::ompl::geometric::PathSimplifier simplifier{si};
+
+  // Flag for user to know if shorten was successful
+  bool shorten_success = false;
+
+  // Set the parameters for termination of simplification process
+  std::chrono::system_clock::time_point const time_before
+      = std::chrono::system_clock::now();
+  std::chrono::system_clock::time_point time_current;
+  std::chrono::duration<double> const time_limit
+      = std::chrono::duration<double>(_timeout);
+  double empty_steps = 0;
+
+  do
+  {
+
+    bool const shortened
+        = simplifier.shortcutPath(path, 1, _maxEmptySteps, 1.0, 0.0);
+    if(shortened) 
+      empty_steps = 0;
+    else
+      empty_steps += 1;
+    
+    time_current = std::chrono::system_clock::now();
+    shorten_success = shorten_success || shortened;
+
+  } while (time_current - time_before <= time_limit
+           && empty_steps <= _maxEmptySteps);
+
+  // Step 4: Convert the simplified geomteric path to AIKIDO untimed trajectory
+  auto returnTraj = toInterpolatedTrajectory(path, _stateSpace, _interpolator);
+
+  // Step 5: Return trajectory and notify user if shortening was successful
+  std::pair<std::unique_ptr<trajectory::Interpolated>, bool> returnPair;
+  return std::make_pair(std::move(returnTraj), shorten_success);
 }
+//=============================================================================
+
+// Following are helper functions. Might want to move them elsewhere for generic
+// use
+
+::ompl::geometric::PathGeometric toOMPLTrajectory(
+    const trajectory::InterpolatedPtr& _interpolatedTraj,
+    ::ompl::base::SpaceInformationPtr _si)
+{
+  auto sspace
+      = ompl_static_pointer_cast<GeometricStateSpace>(_si->getStateSpace());
+
+  ::ompl::geometric::PathGeometric returnPath{std::move(_si)};
+
+  for (size_t idx = 0; idx < _interpolatedTraj->getNumWaypoints(); ++idx)
+  {
+    auto ompl_state = sspace->allocState(_interpolatedTraj->getWaypoint(idx));
+    returnPath.append(ompl_state);
+    sspace->freeState(ompl_state);
+  }
+  return returnPath;
 }
+
+std::unique_ptr<trajectory::Interpolated> toInterpolatedTrajectory(
+    const ::ompl::geometric::PathGeometric& _path,
+    statespace::StateSpacePtr _stateSpace,
```

Comment:
Maybe. It's hard to tell:

- If the compiler inlines the `getStateSpace()` call, then it could potentially be as efficient as passing the value in as an argument. We'd end up with the same number of `shared_ptr` copies either way.
- In this case, `Interpolator` is a pure `virtual` base class, so the call will not be inlined and will incur the penalty of a `virtual `function call (i.e. a `vtable` lookup). There is definitely _some_ overhead for a `virtual` function call, but it's often *much* less than you'd expect with a modern optimizing compiler.
- If we cared a lot about performance, then we might get the best of both worlds by making the type of the `_interpolator` parameter a template argument. A good compiler could then [devirtualize](http://hubicka.blogspot.com/2014/01/devirtualization-in-c-part-1.html) the function, then inline it to achieve the same result as (1), i.e. zero overhead.

If we really wanted to know, we'd have to run a benchmark and, if it's performance is an issue, look at the assembly code the compiler generated and determine why it's slow.

But...the difference is negligible. This is *especially* true since we only call this function once per (very long) planning query. There are also a lot of other optimizations that would get us more bang-for-the-buck.

**TL;DR:** Premature optimization is the root of all evil. Eliminating an unnecessary parameter provides a cleaner API and eliminates an entire class of errors (i.e. mixing different state spaces).